flang/lib/Lower/Allocatable.cpp

```diff
@@ -88,12 +115,257 @@ genAllocatableDeallocate(Fortran::lower::FirOpBuilder &builder,
+    }
+    auto addrVar = box.getMutableProperties().addr;
+    return builder.create<fir::LoadOp>(loc, addrVar);
+  }
+  /// Return {lbound, extent} values read from the MutableBoxValue given
+  /// the dimension.
+  std::pair<mlir::Value, mlir::Value> readShape(int dim) {
+    auto idxTy = builder.getIndexType();
+    if (IRbox) {
+      auto dimVal = builder.createIntegerConstant(loc, idxTy, dim);
+      auto dimInfo = builder.create<fir::BoxDimsOp>(loc, idxTy, idxTy, idxTy,
+                                                    IRbox, dimVal);
+      return {dimInfo.getResult(0), dimInfo.getResult(1)};
+    }
+    const auto &mutableProperties = box.getMutableProperties();
+    auto lb = builder.create<fir::LoadOp>(loc, mutableProperties.lbounds[dim]);
+    auto ext = builder.create<fir::LoadOp>(loc, mutableProperties.extents[dim]);
+    return {lb, ext};
+  }
+
+  /// Return the character length. If the length was not deferred, the value
+  /// that was specified is returned (The mutable fields is not read).
+  mlir::Value readCharacterLenght() {
+    if (box.hasNonDeferredLenParams())
+      return box.nonDeferredLenParams()[0];
+    if (IRbox)
+      return Fortran::lower::CharacterExprHelper{builder, loc}
+          .readLengthFromBox(IRbox);
+    const auto &deferred = box.getMutableProperties().deferredParams;
+    if (deferred.empty())
+      fir::emitFatalError(loc, "allocatable entity has no length property");
+    return builder.create<fir::LoadOp>(loc, deferred[0]);
+  }
+  /// Read all mutable properties. Return the base address.
+  mlir::Value read(mlir::SmallVectorImpl<mlir::Value> &lbounds,
+                   mlir::SmallVectorImpl<mlir::Value> &extents,
+                   mlir::SmallVectorImpl<mlir::Value> &lengths) {
+    auto rank = box.rank();
+    for (decltype(rank) dim = 0; dim < rank; ++dim) {
+      auto [lb, extent] = readShape(dim);
+      lbounds.push_back(lb);
+      extents.push_back(extent);
+    }
+    if (box.isCharacter())
+      lengths.emplace_back(readCharacterLenght());
+    else if (box.isDerived())
+      mlir::emitError(loc,
+                      "TODO: read allocatable or pointer derived type lengths");
+    return readBaseAddress();
+  }
+
+private:
+  Fortran::lower::FirOpBuilder &builder;
+  mlir::Location loc;
+  fir::MutableBoxValue box;
+  mlir::Value IRbox;
+};
+
+/// MutablePropertyWriter allows modifying the properties of a MutableBoxValue.
+class MutablePropertyWriter {
+public:
+  MutablePropertyWriter(Fortran::lower::FirOpBuilder &builder,
+                        mlir::Location loc, const fir::MutableBoxValue &box)
+      : builder{builder}, loc{loc}, box{box} {}
+  /// Update MutableBoxValue with new address, shape and length parameters.
+  /// Extents and lbounds must all have index type.
+  /// lbounds can be empty in which case all ones is assumed.
+  /// Length parameters must be provided for the length parameters that are
+  /// deferred.
```

Comment:
I am not really sure I get the question. Do you mean, how do we skip derived type parameters computation in allocate statement if they are not needed and the expression is pure ?
An allocate statement will look like `allocate(type(t(len_expr)):: t1, t2, t3)`. The length type parameters specified in the allocate statement are not uniquely linked to a single variable being allocated by the allocate statement (`len_expr` applies to both t1, t2, and t3. Maybe the related length parameters is deferred in `t1`, but not in `t2`). So right now, the type spec is lowered at the beginning of allocate statement lowering before looping through the variables being allocated.

So the expression computing the length `len_expr` is always computed, even if it turns out not to be used (because for instance it is not deferred in `t1`, `t2` and `t3` and we are not doing runtime check that `len_expr` is in line with the non-deferred value.). I expect DCE should be able to get rid of it if the expression is pure.

Reply:
Here is some code that compiles with gfortran and ifort.
```fortran
subroutine sub(i)
   integer i
   type u
      integer :: fld1
      complex :: fld2
   end type u
   type t(p1,p2)
      integer, len :: p1,p2
      real :: mem1
      type(u) :: mem2(p1)
      real :: mem3(p2)
   end type t
   type(t(i,:)), allocatable :: arr(:) !declaration
   print *, arr(5:55:10)%mem2(42)%fld2
   print *, arr(5:55:10)%mem3(41)
end
```
The earlier question was about where `i` fits in here on the line `!declaration`.

But this example raises another question in that `!declaration` has a deferred LEN type parameter that isn't specified and these compilers swallow it.

Reply:
OK, so this code is indeed wrong because `arr` is used (in the print) without being un-allocated. Using an unallocated allocatable is forbidden, but Fortran standard does not place a requirement on the compiler to catch it because it cannot be deduced by the compiler in all cases.

Here is what lowering would do in this case:
1. Lower the declaration:
       + Create a MutableBoxValue for `arr`
       + Lower non-deferred parameter `i` (load it), add it to non-deferred parameters list in MutableBoxValue.
       + Allocate a variable for the deferred parameter of `arr`, and place  its address to `MutableBoxValue.mutableProperties.deferredParams`.
       + Allocate variables for the address, the lower and upper bounds and place the variable addresses in `MutableBoxValue.mutableProperties`
       + Set the address variable to `null`
      
2. When lowering the reference to `arr(5:55:10)`, the `MutableBoxValue` will be raised to a `BoxValue` by `genMutableBoxRead`. That is, the bounds and address variables will be read to build the `BoxValue` address and bounds. The length parameters vector of `BoxValue`  will be filled with the non-deferred value of `p1`, and the value loaded from the variable for `p2`.

3.  The rest of `arr(5:55:10)%mem2(42)%fld2` will happen as if `arr` was a `BoxValue`. The resulting address computed for it will be undefined because the bounds and `p2` variables are undefined (since `arr` was never allocated).

The example Fortran code does not actually trigger any `MutablePropertyWriter` use (that is defined in the code where we are currently commenting). `MutablePropertyWriter` is only used on allocation/association (and when returning from functions that may have modified an allocatable that we are not tracking through a descriptor), in all these cases, we will have a value for to provide for all deferred length parameters (and the front-end ensures it).

For instance, in the case of an allocation with a missing deferred type parameter:
```
 type t(p1)
  integer, len :: p1
  real :: x(p1)
 end type
 type(t(:)), allocatable :: y1
 type(t(100)), allocatable :: y2
 allocate(y1) ! NOK
 allocate(y2) ! OK
end
```
f18 will yield an error: "Either type-spec or source-expr must appear in ALLOCATE when allocatable object has a deferred type parameters". ifort also yields an error, and well, gfortran segfaults.